### PR TITLE
mdBook Assets - CSV Approach

### LINKED ISSUE #3841: Generate makefile with mdBook
From https://github.com/JacquesCarette/Drasil/discussions/3835#discussioncomment-9992416.

Generate a makefile alongside the generated mdBook projects so users don't have to remember the mdBook shell commands.

Command to build the book:
```
mdbook build
```

Command to launch and open the server:
```
mdbook serve --open
```

### PR DESCRIPTION
**Pull request by BilalM04**
## Additions

- Generates a CSV file (`assets.csv`) alongside each mdBook project, mapping the actual location of the assets to the location that is able to be rendered locally by mdBook.
- Generates a Makefile along side each mdBook project to build book and launch local server.
- Wrote a script which goes into a project directory and utilizes the CSV file to copy over the assets.
- Modified the main Makefile to run the script before building and/or launching server for the examples.
- Added tracegraphs as a dependency to the mdBook targets in the main Makefile.
- Renamed `Book.hs` to `Config.hs` as it now includes the CSV printer functions.

Closes #3841

### CODE REVIEW

**Review by samm82 — APPROVED** [review]
A minor nitpick and a question for the future. Didn't verify the output (I can if that would be helpful to have a Windows opinion), but this PR looks good!

File: code/scripts/mdBook_assets.sh
Comment (balacij) [inline code/scripts/mdBook_assets.sh:26]: The file looks good to me, but just making sure, did you run `shellcheck` on it?

File: code/scripts/mdBook_assets.sh
Comment (balacij) [inline code/scripts/mdBook_assets.sh:2]: Instead of the file being called `mdbook_assets.sh`, it should be something along the lines of `copy_example_assets.sh` because it can copy other things too.

File: code/scripts/mdBook_assets.sh
Comment (balacij) [inline code/scripts/mdBook_assets.sh:13]: It should be searching for a single, specific file. I think it would be good for us to use a UNIX-style hidden file, e.g., `.drasil-requirements.csv` or something of the sorts.

File: code/stable/dblpend/SRS/mdBook/assets.csv
Comment (balacij) [inline code/stable/dblpend/SRS/mdBook/assets.csv:5]: I wonder if this script should be extended to all of our case studies and used similarly. When I open a fresh build of any project, images don't show, so this would be helpful.

File: code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs
Comment (balacij) [inline code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs:37]: 1. I think we can move a generic "make CSV" function to `drasil-utils` that works on purely String matrices.
2. Making a CSV is a bit more complicated than concatenation with a comma -- you need to escape some characters and use double quotation marks when appropriate.
3. I like it when CSVs have header rows. Do you think we can add them?

File: code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs
Comment (balacij) [inline code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs:54]: If a variable defined in a `where` block only appears once the expression, I think we can just inline it unless the expression is complex or the variable name defined makes the expression more readable.

File: code/stable/hghc/SRS/mdBook/Makefile
Comment (balacij) [inline code/stable/hghc/SRS/mdBook/Makefile:7]: Some of our other generated Makefiles have a watermark on them containing some Drasil information. e.g.,

https://github.com/JacquesCarette/Drasil/blob/e8797118de04997a450458cbfd6e98f881559275/code/stable/dblpend/src/java/Makefile#L1-L5

Do you think we can/should add them here too?

Comment (balacij) [reply code/stable/hghc/SRS/mdBook/Makefile:7]: Note: we don't necessarily need to add Project Name/Purpose, but the Drasil-generated watermark is nice, I think.

**Review by balacij — CHANGES_REQUESTED** [review]

File: code/drasil-gen/lib/Language/Drasil/Generate.hs
Comment (samm82) [reply code/drasil-gen/lib/Language/Drasil/Generate.hs]: Sorry if the "?" seemed passive-aggressive; I meant it more of a "is there a reason we can't do this?" - I wasn't sure if I was missing something 😂 thanks!

File: code/scripts/mdBook_assets.sh
Comment (BilalM04) [reply code/scripts/mdBook_assets.sh]: Yes! ShellCheck doesn't complain.

File: code/scripts/mdBook_assets.sh
Comment (BilalM04) [reply code/scripts/mdBook_assets.sh]: Fixed in https://github.com/JacquesCarette/Drasil/pull/3848/commits/a336d65dc8bfea49f692102e0915753837537b50.

File: code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs
Comment (BilalM04) [reply code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs]: Fixed in https://github.com/JacquesCarette/Drasil/pull/3848/commits/a336d65dc8bfea49f692102e0915753837537b50.

File: code/stable/hghc/SRS/mdBook/Makefile
Comment (BilalM04) [reply code/stable/hghc/SRS/mdBook/Makefile:9]: Added in https://github.com/JacquesCarette/Drasil/pull/3848/commits/798bdf23e25ccb25cb954454f67ca8868d58e588. Also added the watermark for the TeX Makefiles.

File: code/scripts/mdBook_assets.sh
Comment (BilalM04) [reply code/scripts/mdBook_assets.sh]: Done in https://github.com/JacquesCarette/Drasil/pull/3848/commits/8de06be5bd5bb6da9bdc80b552445157292deb44.

File: code/drasil-printers/package.yaml
Comment (BilalM04) [reply code/drasil-printers/package.yaml:18]: Used to import [`System.FilePath`](https://hackage.haskell.org/package/filepath-1.5.3.0/docs/System-FilePath.html).

It is used here:
https://github.com/JacquesCarette/Drasil/blob/798bdf23e25ccb25cb954454f67ca8868d58e588/code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs#L50

It has some other possibly useful functions such as [`takeDirectory`](https://hackage.haskell.org/package/filepath-1.5.3.0/docs/System-FilePath.html#v:takeDirectory), [`takeExtension`](https://hackage.haskell.org/package/filepath-1.5.3.0/docs/System-FilePath.html#v:takeExtension), [`takeBaseName`](https://hackage.haskell.org/package/filepath-1.5.3.0/docs/System-FilePath.html#v:takeBaseName), etc. If you know of any places that could use these functions, I can create a separate issue for that. 

---

And yes, verifying the output on a Windows machine would be nice. Thank you!!

Comment (samm82) [reply code/drasil-printers/package.yaml:18]: The Makefiles seem to be working fine, but the assets still aren't being loaded in the mdBook after running/opening it. Is there another step required to get these copied over? My guess is that it's an issue with how file paths are handled on UNIX vs. Windows systems (maybe making more use of `System.FilePath` could help to make file paths OS-agnostic?)

Along that line, if you have time, creating an issue (at least for investigating how we can be using `System.FilePath` more) would be great!

File: code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs
Comment (BilalM04) [reply code/drasil-printers/lib/Language/Drasil/Markdown/Config.hs]: Added in https://github.com/JacquesCarette/Drasil/pull/3848/commits/d880a81de3b040ee2ac707e358cc3ea209b8904a. I made a new module as I couldn't find an existing one where this would fit.

- I add double quotes around items that contain a `,`, `"`, `\n`, or an empty space.
- I escape double quote characters `"`, using the CSV escape char, which by default is a double quote. 
- Added headings to the CSV file that generates with the mdBook projects.

Please let me know of any other formatting details for CSV files if I missed any.

File: code/drasil-printers/package.yaml
Comment (BilalM04) [reply code/drasil-printers/package.yaml:18]: Could you provide the output log of the Bash script that runs in the Makefile target? On a successful run, there should be no error messages and should be followed immediately by the command that builds the book or launches the server (depending on which target you used):

```
bash "scripts/copy_example_assets.sh" "build/glassbr/SRS/mdBook"
mdbook serve --open "build/glassbr/SRS/mdBook"
```

Comment (samm82) [reply code/drasil-printers/package.yaml:18]: The output of `make projectile_mdBook_server`:
```bash
C:/Program Files/Git/usr/bin/sh.exe "scripts/copy_example_assets.sh" "build//SRS/mdBook"
/usr/bin/sh: line 1: C:/Program: No such file or directory
make: [projectile_mdBook_server] Error 127 (ignored)
mdbook serve --open "build//SRS/mdBook"
2024-07-17 15:55:43 [ERROR] (mdbook::utils): Error: Couldn't open SUMMARY.md in "C:\\Users\\samcr\\Desktop\\Drasil\\code\\build//SRS/mdBook\\src" directory
2024-07-17 15:55:43 [ERROR] (mdbook::utils):    Caused By: The system cannot find the path specified. (os error 3)
make: *** [projectile_mdBook_server] Error 101
```

The output of `make projectile_mdbook_server` (lowercase!):
```bash
C:/Program Files/Git/usr/bin/sh.exe "scripts/copy_example_assets.sh" "build/projectile/SRS/mdBook"
/usr/bin/sh: line 1: C:/Program: No such file or directory
make: [projectile_mdbook_server] Error 127 (ignored)
mdbook serve --open "build/projectile/SRS/mdBook"
2024-07-17 15:58:24 [INFO] (mdbook::book): Book building has started
2024-07-17 15:58:24 [INFO] (mdbook::book): Running the html backend
2024-07-17 15:58:24 [INFO] (mdbook::cmd::serve): Serving on: http://localhost:3000
2024-07-17 15:58:24 [INFO] (mdbook): Opening web browser
2024-07-17 15:58:24 [INFO] (warp::server): Server::run; addr=127.0.0.1:3000
2024-07-17 15:58:24 [INFO] (warp::server): listening on http://127.0.0.1:3000
2024-07-17 15:58:24 [INFO] (mdbook::cmd::watch::poller): Watching for changes...
```

This runs, but doesn't copy over the assets, and seems to be a problem with the fact that my `sh.exe` is located in `C:/Program Files`, which has a space. Would this be on the user to resolve or should we look into handling this on Drasil's end? @smiths

Comment (BilalM04) [reply code/drasil-printers/package.yaml:18]: That's odd, assuming the rest of the Makefile targets work for you (some of which also include the use of scripts), this one should too. Looking at the other targets, the `$(SHELL)` variable in the Makefile are in quotes, where as in mine I was missing the quotes. I just added them in https://github.com/JacquesCarette/Drasil/pull/3848/commits/072ad883f2aa03473c90e944660a084d9b525113, which I hope resolves this issue.

Comment (samm82) [reply code/drasil-printers/package.yaml:18]: Fantastic, it did! 🥳🥳🥳

**Comment by smiths** [comment]
Where are the instructions for using the main `Makefile` to build the mdBook examples, copy over the assets, and start the server?  I can run the examples using the information provided in this issue, but I will forget the details fairly quickly.  We should summarize `make projectile_mdbook_server` etc. somewhere that it is easy to find.  I feel like there is a summary of the main `Makefile` somewhere.  That would be a good location for this information.  (This might have already been done, in which case you can ignore my comment.)  :-)

**Comment by BilalM04** [comment]
@smiths the information for the new `Makefile` targets are included in `make help`. I did find a [wiki page](https://github.com/JacquesCarette/Drasil/wiki/Makefile) that lists the "most commonly used targets".  Should I add it there as well?

**Comment by smiths** [comment]
Thank you @BilalM04!  Yes adding to the wiki would be a great idea.  Where can someone find the possible values for `X` in `X_mdbook_server` (and other commands)?  Would it be possible for `make help` to list the potential values of `X`?  I guess the alternative is for the user to read the `Makefile` to find them.

**Comment by BilalM04** [comment]
User can run `make help_examples`, which is also specified in the intro for the example-specific targets:

```
Example-specific targets where X is the example. Run "make help_examples" to see a list of possible examples:
  X_gen               Generate individual example in HTML and LaTeX format.
  X_diff              Generate individual example and create a comparison against stable in the log folder.
  X_debug             _diff, but in debug mode.
```

**Comment by smiths** [comment]
Great!  Yes, that works well.  Can you please add `make help_examples` to the wiki you mentioned above?

**Review by JacquesCarette — APPROVED** [review]

**Comment by JacquesCarette** [comment]
@samm82 can you re-review? @BilalM04 you might need to merge master and/or rebase for my approval to "work".

**Comment by BilalM04** [comment]
Merged in `main`. I thinks its @balacij that needs to re-review as he requested changes.

**Review by samm82 — APPROVED** [review]

**Review by balacij — APPROVED** [review]
Great work! Thank you, @BilalM04 ! :smile: 

Sorry for holding up the merge! @JacquesCarette it should be good to go from my end now.

**Pull request merged**
